### PR TITLE
Initial draft of policies and guidelines for libcudf usage.

### LINKED ISSUE #4511: [DOC] Discuss and document expected stream synchronization behavior of libcudf functions
**Report needed documentation**

Today, the (a)synchronous behavior of cuDF/libcudf functions is not formally documented (nor has it really been discussed to my knowledge).

The current state of affairs is that no libcudf function explicitly synchronizes before returning. This means that the result buffers from libcudf functions may not immediately be safe to use. 

Thus far, this has been okay because of a few reasons:
- We only use the default stream
- Any operation we do on the result buffer is stream-ordered

This means that for any function:
```
result = cudf::function(...)
```
when `cudf::function` returns, the buffer in `result` may not be safe to immediately use. However, any operations like copying `result` to host or using it in another kernel are all safe because they are stream-ordered. Meaning, any stream ordered operations after `cudf::function` will be en-queued in the stream such that by the time those operations occur, previous operations in the stream will have completed (i.e., `result` is safe to use). 

Where this breaks down:
- Anyone that uses a non-blocking, non-default stream
- Anything that assumes the result buffer is immediately safe to use (like UCX) 

None of this information is really documented anywhere, so we should probably remedy that. 

**Describe the documentation you'd like**

The synchronous behavior of the library is very important to get right. We should have a conversation about what we have now and if it is what we think is correct for the long term future of the library.

Compare to Thrust which takes a more pessimistic approach to synchronization and synchronizes the default stream before returning from functions like reductions or constructing a `device_vector`.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR adds a section to the developer documentation about various libcudf design decisions that affect users. These policies are important for us to document and communicate consistently. I am not sure what the best place for this information is, but I think the developer docs are a good place to start since until we address #11481 we don't have a great way to publish any non-API user-facing libcudf documentation. I've created this draft PR to solicit feedback from other libcudf devs about other policies that we should be documenting in a similar manner. Once everyone is happy with the contents, I would suggest that we merge this into the dev docs for now and then revisit a better place once we've tackled #11481.

Partly addresses #5505, #1781.

Resolves #4511.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (hyperbolic2346) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:360]: This makes it sound like validation would only be a minimal impact. I think the intent here is more that it, at best, is a the impact of the kernel launch, but is cost prohibitive overall.

**Review by hyperbolic2346 — COMMENTED** [review]
Small clarification request.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (davidwendt) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:365]: This extends to output sizes and results as well.

In general there is no validation that an operation will produce a column that fits in the 2GB limit. There are a few minor exceptions where the validation does not require an extra kernel launch.

Also, operations that may produce overflow (integer or floating-point arithmetic) are also not checked. Overflow produces undefined results. For example, integer overflow to negative values is not guaranteed.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (davidwendt) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:375]: There is also the rule that compound columns contain nulls and a valid offset only at the parent level. Children do not have null masks, null counts, or an offset value (i.e. offset > 0 can occur on sliced columns).

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:365]: My intent here was not to provide exhaustive lists of things we will and will not check, but rather to provide some examples of cases to illustrate the difference between introspective and non-introspective operations. I am wary of trying to maintain a complete list that will inevitably go out of date the day after it is published. Do you think it's worth trying to keep a larger list?

Comment (davidwendt) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:365]: I suppose I would like to keep a list of things that keep being brought up (like these two here) so that I can point to this document to say this is by design.

Comment (bdice) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:365]: I agree this does not need to be _exhaustive_, but it would be nice if this policy guide can serve as a mini-FAQ of "why is libcudf designed this way?" I think that the concrete examples @davidwendt mentioned help illustrate the philosophy of the library.

Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:365]: I've added some, and I've reworded this section to reflect that we want to keep a list but make no promises that it's exhaustive.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:367]: Well, the question is whether we should include checking anything around null counts in the list of things that libcudf should validate. If null counts are known a priori then I would argue that we should at least validate cases where we need `null_count == 0`.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (jrhemstad) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:352]:
```suggestion
`libcudf` is designed to provide GPU-accelerated algorithm primitives for solving a wide variety of problems that arise in data science.
Our goal is to enable diverse use cases like Spark or Pandas to benefit from the performance of GPUs. 
As a result, libcudf prioritizes performance and flexibility, which sometimes may come at the cost of convenience.
While we welcome users to use libcudf directly, we design with the expectation that most users will be consuming libcudf through higher-level layers like Spark or cuDF Python. 
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (jrhemstad) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:392]: We synchronize streams all the time. I think the intention was to say "We do not go out of our way to synchronize the stream to ensure the data is immediately available after the function returns". 

```suggestion
## libcudf APIs are not always synchronous
```

**Review by jrhemstad — APPROVED** [review]
Happy to see us writing down some things that have been tribal knowledge for a while now.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:362]:
```suggestion
2. Since libcudf data structures store data on the GPU, any validation incurs _at minimum_ the overhead of a kernel launch, and may in general be prohibitively expensive.
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (bdice) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:377]: Yes — if null counts are always known (on the host?) then checking for nulls would have a similar cost to checking dtypes or input sizes. We just don’t want to incur kernel launches during validation.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:386]:
```suggestion
- For compound columns, nulls should only be present at the level of the parent column. Child columns should not contain nulls.
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:397]:
```suggestion
Work in libcudf occurs on `cudf::get_default_stream().value`, which defaults to the CUDA default stream (stream 0).
```

**Review by bdice — APPROVED** [review]
Only minor comments. Thanks!

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (vyasr) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:377]:
```suggestion
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:388]: Maybe we can resolve this TODO before merging. I don't know exactly what is included in "ensure the above" but in my understanding we want to set the expectation that any/all APIs can _expect_ sanitized inputs with respect to nested nulls, whether or not they currently _rely_ on that behavior.

Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:388]: Ah thanks for spotting that we hadn't addressed this yet. Yeah, that's all that I wanted to get across; even if the comparators don't require it (and as I stated above I think we can get away with it in some cases) we should set that as our API expectation. I'll remove this now.

Comment (vyasr) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:388]:
```suggestion
```

**Comment by vyasr** [comment]
@davidwendt if you are satisfied with how I've addressed your comments feel free to merge this once you approve.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (vyasr) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:352]:
```suggestion
`libcudf` is designed to provide thread-safe, single-GPU accelerated algorithm primitives for solving a wide variety of problems that arise in data science.
APIs are written to execute on the default GPU, which can be controlled by the caller through standard CUDA device APIs or environment variables like `CUDA_VISIBLE_DEVICES`.
Our goal is to enable diverse use cases like Spark or Pandas to benefit from the performance of GPUs, and libcudf relies on these higher-level layers like Spark or Dask to orchestrate multi-GPU tasks.

```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md
Comment (vyasr) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:353]:
```suggestion
To best satisfy these use-cases, libcudf prioritizes performance and flexibility, which sometimes may come at the cost of convenience.
```

**Review by davidwendt — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**
